src/chain/shared_state.rs

```diff
@@ -627,44 +626,58 @@ impl SectionProofChain {
         }
     }
 
-    fn validate_next_block(&self, block: &SectionProofBlock) -> bool {
-        let last = self.last_public_key_info();
-        if block.version() != last.version() + 1 {
-            log_or_panic!(
-                LogLevel::Error,
-                "Invalid next block version - not successor (last: {}, next: {})",
-                last.version(),
-                block.version()
-            );
-            return false;
+    fn validate_next_block(
+        &self,
+        last: &SectionKeyInfo,
+        next: &SectionProofBlock,
+    ) -> Result<(), NextBlockError> {
```

Comment:
I'm not sure why you moved the log or panic outside of this function, it does not seem clearer, and we really should really log or panic for either of the callers.

Reply:
> should really log or panic for either of the callers

I disagree. In `push`, it's we who are calling it so we are fully in control of what we pass in so panicking is appropriate (passing an invalid block is a contract violation), but in `verify` we are calling it on a value that we received from someone else which can contain anything so we need to handle errors gracefully. 

You can also argue that this panics only in tests where we control everything but there seems to be push against `log_or_panic` so I wanted to make the code a bit future-proof.

Reply:
Not 100% sure, so I'll follow what you say.

Reply:
I moved the `log_or_panic` outside of the function so that we can use it for both use cases where the panic is appropriate only in one of them.